File: benchmarks/skynet/skynet.go
// Skynet Benchmark - Go version
//
// Spawns 1,000,000 goroutines in a 10-ary tree structure.
// Each leaf goroutine sends its ID to its parent.
// Parent goroutines sum their 10 children's values.
// The root should return 499999500000 (sum of 0..999999).

package main

import (
	"fmt"
	"time"
)

func skynet(result chan<- int64, num, size int64) {
	if size == 1 {
		result <- num
		return
	}

	// Create channel for children
	children := make(chan int64, 10)
	childSize := size / 10

	// Spawn 10 children
	for i := int64(0); i < 10; i++ {
		go skynet(children, num+i*childSize, childSize)
	}

	// Sum results from children
	var sum int64
	for i := 0; i < 10; i++ {
		sum += <-children
	}

	result <- sum
}

func main() {
	start := time.Now()

	result := make(chan int64)
	go skynet(result, 0, 1000000)

	sum := <-result

	elapsed := time.Since(start)

	fmt.Printf("Result: %d\n", sum)
	fmt.Printf("Time: %d ms\n", elapsed.Milliseconds())
}

File: benchmarks/skynet/skynet.rs
// Skynet Benchmark - Rust version
//
// Spawns 1,000,000 tasks using a thread pool.
// Uses std::thread with work-stealing via recursive subdivision.
// Each leaf returns its ID, parents sum children.
// The root should return 499999500000 (sum of 0..999999).
//
// Note: Unlike Go goroutines, Rust std::thread are OS threads.
// This version uses rayon-style recursive parallelism with a thread pool.
// Build: rustc -O -o skynet_rust skynet.rs

use std::sync::mpsc;
use std::thread;
use std::time::Instant;

// Use a thread pool approach: spawn threads only at higher levels,
// compute leaves sequentially
const PARALLEL_THRESHOLD: i64 = 10000; // Below this, compute sequentially

fn skynet_seq(num: i64, size: i64) -> i64 {
    if size == 1 {
        return num;
    }

    let child_size = size / 10;
    let mut sum = 0i64;
    for i in 0..10 {
        sum += skynet_seq(num + i * child_size, child_size);
    }
    sum
}

fn skynet_par(num: i64, size: i64) -> i64 {
    if size <= PARALLEL_THRESHOLD {
        return skynet_seq(num, size);
    }

    let child_size = size / 10;
    let (tx, rx) = mpsc::channel();

    // Spawn threads for children
    let mut handles = Vec::with_capacity(10);
    for i in 0..10 {
        let tx = tx.clone();
        let child_num = num + i * child_size;
        handles.push(thread::spawn(move || {
            let result = skynet_par(child_num, child_size);
            tx.send(result).unwrap();
        }));
    }
    drop(tx);

    // Sum results
    let mut sum = 0i64;
    for _ in 0..10 {
        sum += rx.recv().unwrap();
    }

    for h in handles {
        h.join().unwrap();
    }

    sum
}

fn main() {
    let start = Instant::now();

    let sum = skynet_par(0, 1_000_000);

    let elapsed = start.elapsed();

    println!("Result: {}", sum);
    println!("Time: {} ms", elapsed.as_millis());
}
